addons/mail_client_extension/controllers/main.py

```diff
@@ -0,0 +1,323 @@
+
+    @http.route('/mail_client_extension/auth/confirm', type='http', auth="user", methods=['POST'])
+    def auth_confirm(self, scope, friendlyname, redirect, info=None, do=None, **kw):
+        """
+        Called by the `app_auth` template. If the user decided to allow the app to access Odoo, a temporary auth code
+        is generated and he is redirected to `redirect` with this code in the URL. It shoudl redirect to the app, and
+        the app should then exchange this auth code for an access token by calling
+        `/mail_client_extension/auth/access_token`.
+        """
+        params = {}
+        if do:
+            name = friendlyname if not info else friendlyname + ' ' + info
+            auth_code = self._generate_auth_code(scope, name)
+            params = {'success': 1, 'auth_code': auth_code}
+        else:
+            params = {'success': 0}
+
+        parsed_redirect = werkzeug.urls.url_parse(redirect)
+        params.update(parsed_redirect.decode_query())
+        parsed_redirect = parsed_redirect.replace(query=werkzeug.urls.url_encode(params))
+        updated_redirect = str(parsed_redirect)
+        return http.redirect_with_hash(updated_redirect)
+
+    # In this case, an exception will be thrown in case of preflight request if only POST is allowed.
+    @http.route('/mail_client_extension/auth/access_token', type='json', csrf=False, auth="none", cors="*", methods=['POST', 'OPTIONS'])
+    def auth_access_token(self, auth_code, **kw):
+        """
+        Called by the external app to exchange an auth code, which is temporary and was passed in a URL, for an
+        access token, which is permanent, and can be used in the `Authorization` header to authorize subsequent requests
+        """
+        is_code_valid, auth_message = self._get_auth_code_data(werkzeug.urls.url_unquote(auth_code))
+        if not is_code_valid:
+            return {"error": "Invalid code"}
+        request.uid = auth_message['uid']
+        scope = auth_message['scope'] if auth_message['scope'] else None # avoid empty string scopes ('')
+        api_key = request.env['res.users.apikeys']._generate(scope, auth_message['name'])
+        return {'access_token': api_key }
+
+    def _get_auth_code_data(self, auth_code):
+        db_secret = request.env['ir.config_parameter'].sudo().get_param('database.secret')
+        data, auth_code_signature = auth_code.split('.')
+        data = base64.b64decode(data)
+        auth_code_signature = base64.b64decode(auth_code_signature)
+        signature = hmac.new(db_secret.encode(), data, hashlib.sha256).hexdigest().encode()
+        if not hmac.compare_digest(auth_code_signature, signature):
+            return False, None
+
+        auth_message = json.loads(data)
+        # Check the expiration
+        if datetime.datetime.utcnow() - datetime.datetime.fromtimestamp(auth_message['timestamp']) > datetime.timedelta(minutes=3):
+            return False, None
+
+        return True, auth_message
```

Comment:
NBD but by definition `is_code_valid` is true IFF `auth_message` is truthy, so seems redundant, caller can just check `if auth_message`.